engines/lab/engine.cpp

```diff
@@ -0,0 +1,1204 @@
+			if (doCloseUp(_closeDataPtr)) {
+				_closeDataPtr = nullptr;
+				mayShowCrumbIndicator();
+				_graphics->screenUpdate();
+			}
+
+			// Sets the current picture properly on the screen
+			if (_mainDisplay)
+				_nextFileName = getPictName(&_closeDataPtr);
+
+			if (_noUpdateDiff) {
+				// Potentially entered another room
+				_roomsFound->inclElement(_roomNum);
+				forceDraw |= (_nextFileName != _curFileName);
+
+				_noUpdateDiff = false;
+				_curFileName = _nextFileName;
+			} else if (_nextFileName != _curFileName) {
+				interfaceOff();
+				// Potentially entered another room
+				_roomsFound->inclElement(_roomNum);
+				_curFileName = _nextFileName;
+
+				if (_closeDataPtr) {
+					switch (_closeDataPtr->_closeUpType) {
+					case SPECIALLOCK:
+						if (_mainDisplay)
+							_tilePuzzle->showCombination(_curFileName.c_str());
+						break;
+					case SPECIALBRICK:
+					case SPECIALBRICKNOMOUSE:
+						if (_mainDisplay)
+							_tilePuzzle->showTile(_curFileName.c_str(), (_closeDataPtr->_closeUpType == SPECIALBRICKNOMOUSE));
+						break;
+					default:
+						_graphics->readPict(_curFileName.c_str(), false);
+						break;
+					}
+				} else
+					_graphics->readPict(_curFileName.c_str(), false);
+
+				drawRoomMessage(curInv, _closeDataPtr);
+				forceDraw = false;
+
+				mayShowCrumbIndicator();
+				_graphics->screenUpdate();
+
+				if (!_followingCrumbs)
+					eatMessages();
+			}
+
+			if (forceDraw) {
+				drawRoomMessage(curInv, _closeDataPtr);
+				forceDraw = false;
+				_graphics->screenUpdate();
+			}
+		}
+
+		// Make sure we check the music at least after every message
+		_music->updateMusic();
+		interfaceOn();
+		IntuiMessage *curMsg = _event->getMsg();
+		if (g_engine->shouldQuit()) {
+			_quitLab = true;
+			return;
+		}
+
+		if (!curMsg) {
+			// Does music load and next animation frame when you've run out of messages
+			gotMessage = false;
+			_music->checkRoomMusic();
+			_music->updateMusic();
+			_anim->diffNextFrame();
+
+			if (_followingCrumbs) {
+				int result = followCrumbs();
+
+				if (result != 0) {
+					uint16 code = 0;
+					switch (result) {
+					case VKEY_UPARROW:
+						code = kButtonForward;
+						break;
+					case VKEY_LTARROW:
+						code = kButtonLeft;
+						break;
+					case VKEY_RTARROW:
+						code = kButtonRight;
+						break;
+					default:
+						break;
+					}
+
+					gotMessage = true;
+					mayShowCrumbIndicator();
+					_graphics->screenUpdate();
+					if (!fromCrumbs(kMessageButtonUp, code, 0, _event->updateAndGetMousePos(), curInv, curMsg, forceDraw, code, actionMode))
+						break;
+				}
+			}
+
+			mayShowCrumbIndicator();
+			_graphics->screenUpdate();
+		} else {
+			gotMessage = true;
+			_followingCrumbs = false;
+			if (!fromCrumbs(curMsg->_msgClass, curMsg->_code, curMsg->_qualifier, curMsg->_mouse, curInv, curMsg, forceDraw, curMsg->_code, actionMode))
+				break;
+		}
+	}
+
+	delete _conditions;
+	delete _roomsFound;
+	delete[] _rooms;
```

Comment:
The LabEngine destructor uses, and deletes, _rooms, so doing it here first is bad. Perhaps all these deletes should be moved to the destructor?